kolibri/plugins/coach/api.py

```diff
@@ -38,48 +38,53 @@ class OptionalPageNumberPagination(pagination.PageNumberPagination):
+class LessonReportPermissions(permissions.BasePermission):
+    """
+    List - check if requester has coach/admin permissions on whole facility.
+    Detail - check if requester has permissions on the Classroom.
+    """
 
-    # check if requesting user has permission for collection or user
     def has_permission(self, request, view):
-        if isinstance(view, LessonReportViewset):
-            report_pk = view.kwargs.get('pk', None)
-            if report_pk is None:
-                # If requesting list view, check if requester has coach/admin permissions on whole facility
-                collection_kind = 'facility'
-                collection_or_user_pk = request.user.facility_id
-            else:
-                # If requesting detail view, only check if requester has permissions on the Classroom
-                collection_kind = 'classroom'
-                collection_or_user_pk = Lesson.objects.get(pk=report_pk).collection.id
-
+        report_pk = view.kwargs.get('pk', None)
+        if report_pk is None:
+            collection_id = request.user.facility_id
         else:
-            if isinstance(view, ClassroomNotificationsViewset):
-                collection_kind = 'classroom'
-            else:
-                collection_kind = view.kwargs.get('collection_kind', 'user')
-            collection_or_user_pk = view.kwargs.get('collection_id', view.kwargs.get('pk'))
+            collection_id = Lesson.objects.get(pk=report_pk).collection.id
 
         allowed_roles = [role_kinds.ADMIN, role_kinds.COACH]
+
         try:
-            if 'user' == collection_kind:
-                return request.user.has_role_for(allowed_roles, FacilityUser.objects.get(pk=collection_or_user_pk))
-            else:
-                return request.user.has_role_for(allowed_roles, Collection.objects.get(pk=collection_or_user_pk))
-        except (FacilityUser.DoesNotExist, Collection.DoesNotExist, ValueError):
+            return request.user.has_role_for(allowed_roles, Collection.objects.get(pk=collection_id))
+        except (Collection.DoesNotExist, ValueError):
             return False
 
 
 class LessonReportViewset(viewsets.ReadOnlyModelViewSet):
-    permission_classes = (permissions.IsAuthenticated, KolibriReportPermissions,)
+    permission_classes = (permissions.IsAuthenticated, LessonReportPermissions,)
     serializer_class = LessonReportSerializer
     queryset = Lesson.objects.all()
 
 
+class ClassroomNotificationsPermissions(permissions.BasePermission):
+    """
+    Allow only users with admin/coach permissions on a collection.
+    """
+
+    def has_permission(self, request, view):
+        collection_id = view.kwargs.get('collection_id')
+
+        allowed_roles = [role_kinds.ADMIN, role_kinds.COACH]
+
+        try:
+            return request.user.has_role_for(allowed_roles, Collection.objects.get(pk=collection_id))
+        except (Collection.DoesNotExist, ValueError):
```

Comment:
If the Collection/Classroom does not exist, does this return a 403 or a 404?

Reply:
Okay, it returns a 403. We probably don't have a special UI for this, but it might be nice to separate both situations (later on). Otherwise if you go to a class that was deleted etc., it will just show a "are you logged in" message (when you actually are logged in).

Reply:
Oh yes, that's right. I wasn't thinking about that much and kept the original `KolibriReportPermissions` logic. But definitely agree with you, it would be nice to return 404 instead and then display another message for such situation on FE. I can open an issue.

Reply:
There's an issue here for error-handling in general #5305

Reply:
Yeah, I think in general, returning `True` from the permissions, and then returning a 404 from the Viewset is a good pattern to follow.